flow/aggregates.py

```diff
@@ -0,0 +1,481 @@
+
+    def __len__(self):
+        return len(self._project)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._project == other._project
+
+    def __hash__(self):
+        blob = self._project.id
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, project):
+        """We have to store self._project when this method is invoked
+        This is because we will then iterate over that project in
+        order to return an aggregates of one.
+        """
+        self._project = project
+
+
+class _MakeAggregates:
+    """This class acts as a funnctor which handles the creation of
+    aggregates using all the jobs in a project.
+
+    :param aggregate:
+        aggregator object associated with this class.
+    :type aggregate:
+        :py:class:`aggregator`
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregate, project):
+        self._aggregate = aggregate
+        self._project = project
+
+    def __call__(self):
+        jobs = list(self._project)
+
+        if self._aggregate._select is not None:
+            jobs = filter(self._aggregate._select, jobs)
+        if self._aggregate._sort_by is not None:
+            jobs = sorted(jobs,
+                          key=lambda job: job.sp[self._aggregate._sort_by],
+                          reverse=self._aggregate._reverse_order)
+        aggregated_jobs = self._aggregate._aggregator_function(jobs)
+        return self._create_nested_aggregate_list(aggregated_jobs)
+
+    def _create_nested_aggregate_list(self, aggregated_jobs):
+        """This method converts the returned subset of jobs as an Iterable
+        from an aggregator_function to an aggregate of jobs as tuple.
+        """
+        nested_aggregates = []
+        aggregate_ids = dict()
+
+        for aggregate in aggregated_jobs:
+            try:
+                filter_aggregate = tuple(filter(self._validate_and_filter_job, aggregate))
+                # Store aggregate in this instance
+                nested_aggregates.append(filter_aggregate)
+                # Store aggregate by their ids in order to search through id
+                aggregate_ids[get_aggregate_id(filter_aggregate)] = filter_aggregate
+            except TypeError:  # aggregate is not iterable
+                ValueError("Invalid aggregator_function provided by the user.")
+
+        return nested_aggregates, aggregate_ids
+
+    def _validate_and_filter_job(self, job):
+        "Validate whether a job is eligible to be a part of an aggregate or not."
+        if job is None:
+            return False
+        elif job in self._project:
+            return True
+        else:
+            raise LookupError(f'The signac job {job.get_id()} not found in {self._project}')
```

Comment:
If we have this class which I generally am in favor of. It should not compose an `Aggregate` object but rather take the values from the object and use them itself.